### ISSUE TITLE
Yet another Action Model

### ISSUE THREAD

**Issue by relu91** [w3c/wot-thing-description#1223]
In today's TD call I proposed an extension of the current hypermedia control proposal for action affordances. After #1222 will be merged you can find it under [proposals folder](./proposals/hypermerdia-control-3). Briefly, the idea is to connect the bits that were already mentioned in different comments and try to describe dynamic action resources with Thing Models and Action Descriptions. This would solve the problem of keep track of ids and how to use them without out-of-band information. Plus it also avoids using dynamic changes in a TD. 

Please use this issue for providing feedback on the proposal.  Next week we'll discuss it further.

**Comment by benfrancis** [comment]
I think this is a clever idea which breaks up the API description using a hypermedia approach.

My biggest concern with this approach is the issue I raised on the call and which you acknowledge in the open questions section, which is how to deal with multiple consumers interacting with a producer simultaneously. i.e.

1. Consumer 1 invokes an action
2. Consumer 2 wants to cancel the action

With the proposed approach consumer 2 would have no way of knowing how to cancel an ongoing action because only consumer 1 has the action description returned when invoking the action.

As you also acknowledge, the proposal doesn't solve the problem of enumerating an action queue, only how to describe follow-up operations on an individual action instance to the consumer which invoked it.

**Comment by egekorkan** [comment]
I don't see the proposal somehow? See screenshot below
![Screenshot_2021-09-14-08-16-44-113_org.mozilla.firefox.jpg](https://user-images.githubusercontent.com/20195407/133199067-9543a8ac-5c3b-470d-b142-008208912ea1.jpg)

**Comment by danielpeintner** [comment]
PR #1222 still needs to be merged to become visible

**Comment by benfrancis** [comment]
I've been thinking some more about this proposal and I've concluded that it's unfortunately solving the wrong problem. The problem I think it solves is how to describe the API to carry out follow-up operations on an ongoing action, but what it doesn't solve is how to describe a queue of actions (e.g. there's no way for a Consumer to get a list of ongoing/queued actions).

The kind of use case where this is useful is one where you have a queryable and cancelable action, but which only supports running/queueing one instance of the action at a time. But this use case can probably be solved more simply with a standard static Thing Description, e.g. 

```json
{
  "@context": "https://www.w3.org/2019/wot/td/v1",
  "id": "urn:ex:thing",
  "title": "Web Lamp",
  "securityDefinitions": {
    "basic_sc": {
      "scheme": "basic",
      "in": "header"
    }
  },
  "actions": {
    "fade": {
      "input": {
        "type": "object",
        "properties": {
          "level": {
            "type": "integer",
            "minimum": 0,
            "maximum": 100
          },
          "duration": {
            "type": "integer",
            "minimum": 0,
            "unit": "milliseconds"
          }
        }
      },
      "output": {},
      "forms": [
        {
          "href": "/actions/fade",
          "op": "invokeaction",
          "htv:methodName": "POST",
          "contentType": "application/json",
          "response": {
            "schema": "actionStatus"
          }
        },
        {
          "href": "/actions/fade",
          "op": "queryaction",
          "htv:methodName": "GET",
          "response": {
            "contentType": "application/json",
            "schema": "actionStatus"
          }
        },
        {
          "href": "/actions/fade",
          "op": "cancelaction",
          "htv:methodName": "DELETE", // Not sure DELETE is the right verb here
          "response": {
            "contentType": "application/json",
            "schema": "actionStatus"
          }
        }
      ]
    }
  },
  "forms": [
    {
      "href": "/actions",
      "op": "queryallactions",
      "htv:methodName": "GET"
    }
  ],
  "schemaDefinitions": {
    "actionStatus": {
      "type": "object",
      "properties": {
        "active": {
          "type": "boolean"
        },
        "output": {},
        "error": {
          "type": "object"
        }
      },
      "required": [
        "active"
      ]
    }
  }
}
```

Maybe that's the most we can realistically support with a static Thing Description, and anything more complex would have to be described out-of-band (e.g. in a profile). See https://github.com/w3c/wot-thing-description/issues/302#issuecomment-919076095

**Comment by egekorkan** [comment]
I sort of like this approach, agree with Ben that it is not solving the queue problem. Some specific comments:
1. I am not sure what kind of brownfield devices can be covered. If the linking to the AM TM (Action Model TM) is not already done, which is the case for a brownfield device, I don't know how this can be linked unless these TMs are available somewhere else than the device itself.
2. Should the consumers get AM TMs proactively? If yes, it implies quite some change to consumer code style. I would need to consume TMs now.
3. I am not sure how the id tracking problem is resolved. Now I would need to track TMs per action request that where TMs have the ids in them. 
4. There is small mentioning of tagging certain affordances with tags that we predefine. This sounds interesting for further exploration.

**Comment by lu-zero** [comment]
@relu91 could you please update the link to the proposal? Is it still relevant or this concern has been already subsumed by other open issues regarding `action` queues management?
